### PR TITLE
Improve subsidy handling

### PR DESCRIPTION
**Pull request by ckoopmann**
1. Pull in subsidy instead of transfering in separate tx (requires approval)
2. Return net subsidy for use in simulation (callStatic)
3. Add min / maxSubsidy for additional slippage protection

### CODE REVIEW

File: contracts/adapters/AaveMigrationExtension.sol [hunk 1]
```diff
@@ -575,4 +554,18 @@ contract AaveMigrationExtension is BaseExtension, FlashLoanSimpleReceiverBase, I
...
+
+    function _checkAndRepaySubsidy(uint256 _maxSubsidy, uint256 _minWrappedSetTokenPosition) internal returns (uint256, uint256){
+        uint256 underlyingTokenBalance = underlyingToken.balanceOf(address(this));
+        uint256 subsidyAmount = _maxSubsidy.sub(underlyingTokenBalance); // Assumes 0 weth balance in this contract prior to tx
+        if (underlyingTokenBalance > 0) {
+            underlyingToken.transfer(msg.sender, underlyingTokenBalance);
+        }
+        uint256 wrappedSetTokenPositionAfter = uint256(setToken.getDefaultPositionRealUnit(address(wrappedSetToken)));
+        if(wrappedSetTokenPositionAfter< _minWrappedSetTokenPosition){
+            revert("MigrationExtension: Insufficient wrappedSetTokenPosition");
```
Comment (pblivin0x) [inline contracts/adapters/AaveMigrationExtension.sol:566]: this is already checked in the `TradeModule.trade()` call with  `_minReceiveQuantity`  https://github.com/IndexCoop/index-protocol/blob/9e6e7d865add19ebfa4735b9743ebeb3b5029533/contracts/protocol/modules/v1/TradeModule.sol#L122

Comment (pblivin0x) [reply contracts/adapters/AaveMigrationExtension.sol]: removed this check here [648211a](https://github.com/IndexCoop/index-coop-smart-contracts/pull/163/commits/648211a33a0d63b307e9eb88f81d2ec76c4fd503)

Comment (edkim) [reply contracts/adapters/AaveMigrationExtension.sol]: Agree that it looks fine to remove this extra check.

**Review by edkim — APPROVED** [review]
LGTM!

File: contracts/adapters/AaveMigrationExtension.sol [hunk 2]
```diff
@@ -255,46 +255,25 @@ contract AaveMigrationExtension is BaseExtension, FlashLoanSimpleReceiverBase, I
...
-        uint256 _underlyingRedeemLiquidityMinAmount,
-        uint256 _wrappedSetTokenRedeemLiquidityMinAmount
+        uint256 _maxSubsidy
     )
         external
         onlyOperator
+        returns (uint256 subsidyAmount)
     {
+        // Subsidize the migration
+        underlyingToken.transferFrom(msg.sender, address(this), _maxSubsidy);
```
Comment (pblivin0x) [inline contracts/adapters/AaveMigrationExtension.sol:273]: add support for `_maxSubsidy = 0`

Comment (pblivin0x) [reply contracts/adapters/AaveMigrationExtension.sol]: added support here [eca0526](https://github.com/IndexCoop/index-coop-smart-contracts/pull/163/commits/eca05268f6f9b903ec9c472dbde31cf00befb54c)

**Review by edkim — APPROVED** [review]

**Pull request merged**